rust/src/interpreter/gas.rs

```diff
@@ -0,0 +1,68 @@
+use evmc_vm::{AccessStatus, Address, ExecutionContext, Revision, StatusCode};
+
+use crate::{interpreter::utils::word_size, types::u256};
+
+#[inline(always)]
+pub(super) fn consume_gas(gas_left: &mut u64, gas: u64) -> Result<(), StatusCode> {
+    if *gas_left < gas {
+        return Err(StatusCode::EVMC_OUT_OF_GAS);
+    }
+    *gas_left -= gas;
+    Ok(())
+}
+
+#[inline(always)]
+pub(super) fn consume_positive_value_cost(
+    value: &u256,
+    gas_left: &mut u64,
+) -> Result<(), StatusCode> {
+    if *value != u256::ZERO {
+        consume_gas(gas_left, 9000)?;
+    }
+    Ok(())
+}
+
+#[inline(always)]
+pub(super) fn consume_value_to_empty_account_cost(
+    value: &u256,
+    addr: &Address,
+    context: &mut ExecutionContext,
+    gas_left: &mut u64,
+) -> Result<(), StatusCode> {
+    if *value != u256::ZERO && !context.account_exists(addr) {
+        consume_gas(gas_left, 25000)?;
+    }
+    Ok(())
+}
```

Comment:
I am curious about the code using these two functions ... they seem to be fragments that should be an integral part of some enclosing code -- something computing the cost for a value transfer and only when this is complete, call `consume_gas`.

Could it be that this function, together with the calling code, could be merged into a single `charge_value_transfer_costs` function?

Reply:
`consume_positive_value_cost` gets called for `Call` and `CallCode` while `consume_value_to_empty_account_cost` only gets called for the former. When putting both together there would be an additional argument necessary to indicate if the value to empty account cost applies or not. By keeping them as separate functions it is more clear which costs get subtracted and they correspond 1:1 to the names in [evm.codes](https://www.evm.codes/).